$OpenBSD: patch-src_filelist_c,v 1.3 2014/04/15 13:02:21 dcoppa Exp $

commit 5f210b5b2b0b150acf8f4b5890c9811cd65e2aa9
Author: Michael Vorburger <mike@vorburger.ch>
Date:   Sun Dec 29 22:34:11 2013 +0100

Fix for segfault due to erroneous free() in case of scandir failed
to scan directory

--- src/filelist.c.orig	Fri Feb 28 18:23:20 2014
+++ src/filelist.c	Tue Apr 15 14:40:39 2014
@@ -263,27 +263,27 @@ void add_file_to_filelist_recursively(char *origpath, 
 			default:
 				weprintf("Failed to scan directory %s:", path);
 			}
-		}
+		} else {
+			for (cnt = 0; cnt < n; cnt++) {
+				if (strcmp(de[cnt]->d_name, ".")
+						&& strcmp(de[cnt]->d_name, "..")) {
+					char *newfile;
 
-		for (cnt = 0; cnt < n; cnt++) {
-			if (strcmp(de[cnt]->d_name, ".")
-					&& strcmp(de[cnt]->d_name, "..")) {
-				char *newfile;
+					newfile = estrjoin("", path, "/", de[cnt]->d_name, NULL);
 
-				newfile = estrjoin("", path, "/", de[cnt]->d_name, NULL);
+					/* This ensures we go down one level even if not fully recursive
+					   - this way "feh some_dir" expands to some_dir's contents */
+					if (opt.recursive)
+						add_file_to_filelist_recursively(newfile, FILELIST_CONTINUE);
+					else
+						add_file_to_filelist_recursively(newfile, FILELIST_LAST);
 
-				/* This ensures we go down one level even if not fully recursive
-				   - this way "feh some_dir" expands to some_dir's contents */
-				if (opt.recursive)
-					add_file_to_filelist_recursively(newfile, FILELIST_CONTINUE);
-				else
-					add_file_to_filelist_recursively(newfile, FILELIST_LAST);
-
-				free(newfile);
+					free(newfile);
+				}
+				free(de[cnt]);
 			}
-			free(de[cnt]);
+			free(de);
 		}
-		free(de);
 		closedir(dir);
 	} else if (S_ISREG(st.st_mode)) {
 		D(("Adding regular file %s to filelist\n", path));
